File: 백준/Silver/2178. 미로 탐색/미로 탐색.cc
#include <iostream>
#include <queue>
#include <utility>
#define INF 123456789
using namespace std;
int n, m;
int dxs[4] = { 1, -1, 0, 0 }, dys[4] = { 0, 0, 1, -1 };
char g[101][101];
int maze[101][101];
bool inRange(int a, int b) { return 0 <= a && a < m && 0 <= b && b < n; }
void bfs() {
	queue<pair<int, int>> q;
	q.push(make_pair(0, 0));
	maze[0][0] = 0;
	while (!q.empty()) {
		int x = q.front().first, y = q.front().second;
		q.pop();
		for (int dir = 0; dir < 4; dir++) {
			int nx = x + dxs[dir], ny = y + dys[dir];
			if (inRange(nx, ny) && g[ny][nx] == '1' && maze[y][x] + 1 < maze[ny][nx] ) {
				q.push(make_pair(nx, ny));
				maze[ny][nx] = maze[y][x] + 1;
			}
		}
	}

}
int main() {
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> g[i][j];
			maze[i][j] = INF;
		}
	}
	bfs();
	cout << maze[n - 1][m - 1] + 1;
	return 0;
}

File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque
dxs, dys = [1, -1, 0, 0], [0, 0, 1, -1]
n, m = map(int, sys.stdin.readline().rstrip().split())
g = [list(map(str, sys.stdin.readline().rstrip())) for _ in range(n)]
d = [[123456789 for _ in range(m)] for __ in range(n)]


def in_range(a, b):
    return 0 <= a < m and 0 <= b < n


def bfs():
    dq = deque()
    dq.appendleft([0, 0, 0])
    while dq:
        x, y, length = dq.pop()
        for dx, dy in zip(dxs, dys):
            nx, ny = x + dx, y + dy
            if in_range(nx, ny) and g[ny][nx] == '1' and d[ny][nx] > length + 1:
                dq.appendleft([nx, ny, length + 1])
                d[ny][nx] = length + 1


bfs()
print(d[n - 1][m - 1] + 1)
